internal/ip/ip.go

```diff
@@ -109,25 +139,58 @@ func getSendSocket(af int) (int, error) {
 	return fd, nil
 }
 
-func getRecvSource(listenPort uint32, intf string) (gopacket.PacketDataSource, error) {
-	// Filter for tcp packets coming into ListenPort, http, or HTTPS and contain a SYN flag
-	filter := fmt.Sprintf("(dst port %d or src port %d or src port %d) and (tcp[13] & 2 != 0)",
-		listenPort,
-		d.PortHTTP,
-		d.PortHTTPS)
+func getBPFFilter(ipHeaderOffset uint32, listenPort uint32) ([]bpf.RawInstruction, error) {
+	// The Arachne BPF Filter reads values starting from the TCP Header by adding ipHeaderOffset to all
+	// offsets. It filters for packets of destination port equal to listenPort, or src port equal to HTTP or HTTPS ports
+	// and for packets containing a TCP SYN flag (SYN, or SYN+ACK packets)
+	return bpf.Assemble([]bpf.Instruction{
+		bpf.LoadAbsolute{Off: ipHeaderOffset + 2, Size: 2},              // Starting from TCP Header, load DstPort (2nd word)
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: listenPort, SkipTrue: 3},   // Return packet if DstPort is listen Port
+		bpf.LoadAbsolute{Off: ipHeaderOffset, Size: 2},                  // Starting from TCP Header, load SrcPort (1st word)
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: d.PortHTTP, SkipTrue: 1},   // Return packet if SrcPort is HTTP Port
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: d.PortHTTPS, SkipFalse: 2}, // Discard packet if not HTTPS
+		bpf.LoadAbsolute{Off: ipHeaderOffset + 13, Size: 1},             // Starting from TCP Header, load Flags byte (not including NS bit)
+		bpf.JumpIf{Cond: bpf.JumpBitsSet, Val: 2, SkipTrue: 1},          // AND Flags byte with 00000010 (SYN), and drop packet if 0
+		bpf.RetConstant{Val: 0},                                         // Drop packet
+		bpf.RetConstant{Val: 4096},                                      // Return up to 4096 bytes from packet
+	})
+}
+
+func getRecvSource(af int, listenPort uint32, intf string, logger *log.Logger) (recvSource, error) {
+	var (
+		rs             recvSource
+		ipHeaderOffset uint32
+	)
 
-	handle, err := pcap.OpenLive(intf, d.PcapMaxSnapLen, false, pcap.BlockForever)
+	fd, err := syscall.Socket(af, syscall.SOCK_RAW, syscall.IPPROTO_TCP)
 	if err != nil {
-		return nil, err
+		return rs, err
 	}
 
-	if err = handle.SetBPFFilter(filter); err != nil {
-		return nil, err
+	if err = bindToDevice(fd, intf); err != nil {
+		return rs, err
 	}
 
-	handle.SetDirection(pcap.DirectionIn)
+	rs.fd = fd
+
+	switch af {
+	case d.AfInet:
+		ipHeaderOffset = ip4HeaderLength
+	case d.AfInet6:
+		ipHeaderOffset = ip6HeaderLength
+	}
+	if filter, err := getBPFFilter(ipHeaderOffset, uint32(listenPort)); err != nil {
+		logger.Warn("Failed to compile BPF Filter", zap.Error(err))
+	} else {
```

Comment:
go idiom: early return, remove the else

https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow